### PR TITLE
Feature ctxmng: Repository context manager

### PR DESCRIPTION
**Pull request by enkore**
From https://github.com/borgbackup/borg/issues/562#issuecomment-198576577

Not sure if we should do `@drop_unused_args`, avoids some repetition, but could be a bit "opaque".

Looks good to me. I don't see any ignored (raised in __del__) exceptions in ArchiverTestCase anymore, and only one in RemoteArchiver (test_remote_repo_restrict_to_path, I think it might have something to do with that monkey patching). This should essentially eliminate stale locks in practice. (Except for SIGTERM).

### CODE REVIEW

File: borg/archiver.py [hunk 1]
```diff
@@ -54,20 +55,69 @@ def __call__(self, parser, ns, values, option):
...
+
+
+def with_cache(method):
+    @functools.wraps(method)
+    def wrapper(self, args, repository, key, manifest, **kwargs):
+        with Cache(repository, key, manifest, do_files=getattr(args, 'cache_files', False),
+                   lock_wait=self.lock_wait) as cache:
+            return method(self, args, repository=repository, key=key, manifest=manifest, cache=cache, **kwargs)
+    return wrapper
+
```
Comment (ThomasWaldmann) [inline borg/archiver.py]: quite a lot of magic and additional code, just to save 1 indentation level.

File: borg/remote.py [hunk 2]
```diff
@@ -347,7 +360,7 @@ def load_key(self):
     def break_lock(self):
         return self.call('break_lock')
 
-    def close(self):
+    def _close(self):
```
Comment (ThomasWaldmann) [inline borg/remote.py]: why did you rename? to point out that one should use "with"?

File: borg/archiver.py (hunk 1 above)
Comment (enkore) [inline borg/archiver.py]: I don't see much magic in the with_* decorators, but I agree that `@drop_unused_args` might be too opaque to actually use it outside a Python Obfuscated Code Contest (well, not that bad, but you get the idea).

File: borg/remote.py (hunk 2 above)
Comment (enkore) [inline borg/remote.py]: Yep, _close would now be a private method of Repository, which generally shouldn't be used from client code.

File: borg/archiver.py (hunk 1 above)
Comment (enkore) [inline borg/archiver.py]: I dropped @drop_unused_args, @with_cache was also removed.

Rationale: with_cache and with_repository both operated on the same logical level (repository wide), while with_archive operates on another, smaller level (archive level).

**Comment by enkore** [comment]
I reverted the renaming of _close to avoid breaking the RPC protocol. I tested it manually with borg 1.0 using this as the server, and vice versa, for list|extract|create. Is there a way to do automatic RPC compatibility tests between versions that I overlooked? If there isn't, we should devise one, because manually testing this takes a good while and is of course unreliable like all manual testing (and doesn't use the large test suite we have).

I also extended one test case for borg delete slightly, so that Codecov doesn't complain ;)

File: borg/archiver.py [hunk 3]
```diff
@@ -1,5 +1,6 @@
 from binascii import hexlify, unhexlify
 from datetime import datetime
+from inspect import signature
```
Comment (ThomasWaldmann) [inline borg/archiver.py]: not used any more? (good :)

File: borg/archiver.py [hunk 4]
```diff
@@ -40,18 +41,56 @@
...
+    if isinstance(str_or_bool, str):
+        return getattr(args, str_or_bool)
+    return str_or_bool
 
-    toggle flags are in the form of ``--foo``, ``--no-foo``.
 
-    the ``--no-foo`` argument still needs to be passed to the
-    ``add_argument()`` call, but it simplifies the ``--no``
-    detection.
+def with_repository(fake=False, create=False, lock=True, exclusive=False, manifest=True, cache=False):
```
Comment (ThomasWaldmann) [inline borg/archiver.py]: is there a specific reason why some items default to True, other to False?

might be easier to say "if you want <item>, you need item=True" (and default to False).

Comment (enkore) [inline borg/archiver.py]: The defaults are what is used most commonly:

- 8 commands use the defaults (cache=False, manifest=True, exclusive=False)
- 5 have manifest=False (init, check and all do_debug stuff)
- 4 use cache=True
- 4 use exclusive=True

File: borg/archiver.py (hunk 3 above)
Comment (enkore) [inline borg/archiver.py]: Yep :)

File: borg/remote.py [hunk 5]
```diff
@@ -164,11 +167,21 @@ def __init__(self, location, create=False, lock_wait=None, lock=True, args=None)
         self.id = self.call('open', location.path, create, lock_wait, lock)
 
     def __del__(self):
-        self.close()
+        if self.p:
+            self.close()
+            assert False, "cleanup happened in Repository.__del__"
```
Comment (ThomasWaldmann) [inline borg/remote.py]: this is just for the case, right - or are you still seeing this?

File: borg/repository.py [hunk 6]
```diff
@@ -59,16 +59,31 @@ def __init__(self, path, create=False, exclusive=False, lock_wait=None, lock=Tru
...
+        self.lock_wait = lock_wait
+        self.do_lock = lock
+        self.do_create = create
+        self.exclusive = exclusive
 
     def __del__(self):
-        self.close()
+        if self.lock:
+            self.close()
+            assert False, "cleanup happened in Repository.__del__"
```
Comment (ThomasWaldmann) [inline borg/repository.py]: same q. here?

File: borg/testsuite/archiver.py [hunk 7]
```diff
@@ -1122,8 +1129,9 @@ def setUp(self):
 
     def open_archive(self, name):
         repository = Repository(self.repository_path)
-        manifest, key = Manifest.load(repository)
-        archive = Archive(repository, key, manifest, name)
+        with repository:
+            manifest, key = Manifest.load(repository)
+            archive = Archive(repository, key, manifest, name)
         return archive, repository
```
Comment (ThomasWaldmann) [inline borg/testsuite/archiver.py]: hmm, if you use "with repository" above, it will return a closed repository here.

File: borg/testsuite/key.py [hunk 8]
```diff
@@ -4,11 +4,14 @@
 import tempfile
 from binascii import hexlify, unhexlify
 
+from borg.logger import setup_logging
 from ..crypto import bytes_to_long, num_aes_blocks
 from ..key import PlaintextKey, PassphraseKey, KeyfileKey
 from ..helpers import Location
 from . import BaseTestCase
 
+setup_logging()
```
Comment (ThomasWaldmann) [inline borg/testsuite/key.py]: hmm?

File: borg/remote.py (hunk 5 above)
Comment (enkore) [inline borg/remote.py]: For me all appearances of `__del__` cleanup are gone in the testsuite except the one in test_remote_repo_restrict_to_path (which I think is caused by `patch.object` basically overriding the context manager implemented in Repository, i.e. the patched object is bound instead of the Repository).

File: borg/testsuite/repository.py [hunk 9]
```diff
@@ -100,13 +100,15 @@ def test_single_kind_transactions(self):
         self.repository.close()
         # replace
         self.repository = self.open()
-        self.repository.put(b'00000000000000000000000000000000', b'bar')
-        self.repository.commit()
-        self.repository.close()
+        with self.repository:
+            self.repository.put(b'00000000000000000000000000000000', b'bar')
+            self.repository.commit()
+            self.repository.close()
```
Comment (ThomasWaldmann) [inline borg/testsuite/repository.py]: with ... AND .close() ?

File: borg/testsuite/archiver.py (hunk 7 above)
Comment (enkore) [inline borg/testsuite/archiver.py]: True, the later bind would still fix this, but 'open's aren't exactly cheap. I'll patch it.

File: borg/testsuite/key.py (hunk 8 above)
Comment (enkore) [inline borg/testsuite/key.py]: Got a few failures due to un-initialized loggers when running this one alone without running the entire suite. When running the test suite the archiver test cases will do this.

Maybe put it in __init__ of the test suite instead, to ensure that the loggers are always there for all tests?

Comment (ThomasWaldmann) [inline borg/testsuite/key.py]: yes, i'ld rather have it in test init than at module import time.

File: borg/testsuite/repository.py (hunk 9 above)
Comment (enkore) [inline borg/testsuite/repository.py]: Yeah, that's superfluous. close() is idempotent, so won't blow up either, but it doesn't belong there.

File: borg/testsuite/key.py (hunk 8 above)
Comment (enkore) [inline borg/testsuite/key.py]: Yep `__init__` works too, so I'll put it there instead.

File: borg/remote.py (hunk 5 above)
Comment (enkore) [inline borg/remote.py]: Ah, of course, in the tests tearDown performs "hard" tear downs and just rmtree()s the repository. That probably raced with `__del__` and lead to the locks being deleted before del had a change to clean them up.

**Comment by enkore** [comment]
I guess technically codecov isn't quite right that only 85 % hit, it's just that the llfuse tests don't run on travis... :D

**Comment by ThomasWaldmann** [comment]
Hmm, it's a bit hard to review if you immediately collapse your new changes into your old ones.

**Comment by enkore** [comment]
The only thing changed other than the comment-diffs above is the new changes in upgrader.py, which where in a stash I forgot previously. ([diff](https://gist.github.com/enkore/2175f087dd41ec353d92)). Next time I'll use squash-before-merge approach for larger changeset again
